src/lang/checkers/Cashflow.ml

```diff
@@ -0,0 +1,1313 @@
+      match e with
+      | Literal _ ->
+          (* No need to deduce tag from type. 
+             If the literal is a number, then nothing can be deduced, 
+             and if it is not a number, the tag of relevant variables 
+             will be deduced from their usage. *)
+          (e, tag, field_env, local_env, false)
+      | Var i ->
+          let new_i_tag = lub (lookup_var_tag2 i local_env field_env) in
+          let new_i = update_id_tag i new_i_tag in
+          let (new_local_env, new_field_env) = update_var_tag2 i new_i_tag local_env field_env in
+          (Var new_i, new_i_tag, new_field_env, new_local_env, new_i_tag <> (get_id_tag i))
+      | Fun (arg, t, body) ->
+          (* Using Map tag to represent functions as well as maps *)
+          let body_expected_tag =
+            match expected_tag with
+            | Map x -> x
+            | NoInfo -> NoInfo
+            | _     -> Inconsistent in
+          let body_local_env =
+            AssocDictionary.insert (get_id arg) (get_id_tag arg) local_env in
+          let ((_, (new_body_tag, _)) as new_body, res_field_env, res_body_local_env, body_changes) =
+            cf_tag_expr body body_expected_tag field_env body_local_env in
+          let res_arg_tag = lookup_var_tag arg res_body_local_env in
+          let new_local_env = AssocDictionary.remove (get_id arg) res_body_local_env in
+          (Fun (update_id_tag arg res_arg_tag, t, new_body),
+           Map new_body_tag,
+           res_field_env,
+           new_local_env,
+           body_changes || (get_id_tag arg <> res_arg_tag))
+      | App (f, args) ->
+          let new_args = List.map (fun arg -> update_id_tag arg (lookup_var_tag2 arg local_env field_env)) args in
+          let (args_changes, _) =
+            List.fold_left
+              (fun (acc_changes, new_args_acc) arg ->
+                 match new_args_acc with
+                 | [] -> (false, [])
+                 | x :: rest -> (acc_changes || (get_id_tag x) <> (get_id_tag arg), rest))
+              (false, new_args) args in
+          let f_tag = lub_tags (lookup_var_tag2 f local_env field_env) (Map expected_tag) in
+          let new_f = update_id_tag f f_tag in
+          let (new_local_env, new_field_env) = update_var_tag2 f f_tag local_env field_env in
+          let new_e_tag = 
+            match f_tag with
+            | Map t -> t
+            | NoInfo -> NoInfo
+            | _     -> Inconsistent in
+          (App (new_f, new_args),
+           new_e_tag,
+           new_field_env,
+           new_local_env,
+           args_changes || f_tag <> get_id_tag f)
+      | Builtin (f, args) ->
+          let args_tags = List.map (fun arg -> lookup_var_tag2 arg local_env field_env) args in
+          let (res_tag, args_tags_usage) = builtin_signature f expected_tag args_tags in
+          let final_args_tags =
+            List.map2
+              (fun arg_tag arg_tag_usage -> lub_tags arg_tag arg_tag_usage)
```

Comment:
Is taking an lub here necessary? considering that `builtin_signature` already does it when computing `args_tags_usage`.

Reply:
Well spotted. I have removed it.